pkg/api/merge.go

```diff
@@ -0,0 +1,368 @@
+package api
+
+// TODO: needs unit tests
+
+// Notes: shortcomings of the Azure-style API include:
+// 1. There is no mechanism to delete an item from a slice or map without
+// implementing a separate CRUD endpoint.
+// 2. There is no way to set a plain value to its zero value (e.g. bool to
+// false, int to 0, string to "") without storing it as a pointer in the API.
+
+// At the moment Config isn't merged.  Perhaps what we actually need is a
+// merging converter f(*ext, *int) *int.  Perhaps this would mean we could get
+// rid of some pointers in our internal type.
+
+func Merge(complete, partial *OpenShiftManagedCluster) *OpenShiftManagedCluster {
+	complete = complete.DeepCopy() // required because Config isn't merged
+
+	return mergeOpenShiftManagedCluster(complete, partial)
+}
+
+func mergeOpenShiftManagedCluster(complete, partial *OpenShiftManagedCluster) *OpenShiftManagedCluster {
+	switch {
+	case complete == nil && partial == nil:
+		return nil
+	case complete == nil:
+		complete = &OpenShiftManagedCluster{}
+	case partial == nil:
+		partial = &OpenShiftManagedCluster{}
+	}
+
+	return &OpenShiftManagedCluster{
+		ID:         mergeString(complete.ID, partial.ID),
+		Location:   mergeString(complete.Location, partial.Location),
+		Name:       mergeString(complete.Name, partial.Name),
+		Plan:       mergeResourcePurchasePlan(complete.Plan, partial.Plan),
+		Tags:       mergeTags(complete.Tags, partial.Tags),
+		Type:       mergeString(complete.Type, partial.Type),
+		Properties: mergeProperties(complete.Properties, partial.Properties),
+		Config:     complete.Config, // NOTE: Config isn't merged
+	}
+}
+
+func mergeString(complete, partial string) string {
+	if partial == "" {
+		return complete
+	}
+	return partial
+}
+
+func mergeResourcePurchasePlan(complete, partial *ResourcePurchasePlan) *ResourcePurchasePlan {
+	switch {
+	case complete == nil && partial == nil:
+		return nil
+	case complete == nil:
+		complete = &ResourcePurchasePlan{}
+	case partial == nil:
+		partial = &ResourcePurchasePlan{}
+	}
+
+	return &ResourcePurchasePlan{
+		Name:          mergeString(complete.Name, partial.Name),
+		Product:       mergeString(complete.Product, partial.Product),
+		PromotionCode: mergeString(complete.PromotionCode, partial.PromotionCode),
+		Publisher:     mergeString(complete.Publisher, partial.Publisher),
+	}
+}
+
+func mergeTags(complete, partial map[string]string) map[string]string {
+	if complete == nil && partial == nil {
+		return nil
+	}
+
+	// TODO: is this the semantic that we want for mergeTags?
```

Comment:
Is there any use cases where someone would want to remove a tag? I can't think of any.

Reply:
I'm sure there are.  Will ask the MSFT team about this.